File: LLD/InterviewExamples/RuleEngine/lib/rules/impl/MaximumAmountRule.java
package lib.rules.impl;

import java.util.Optional;

import lib.models.Expense;
import lib.rules.ExpenseRule;
import lib.rules.Violation;
public class MaximumAmountRule implements ExpenseRule{
    
    private final double maxAmount;

    public MaximumAmountRule(double maxAmount) {
        this.maxAmount = maxAmount;
    }

    @Override
    public Optional<Violation> check(Expense e) {
        if(e.getAmountUsd() > maxAmount) {
            return Optional.of(Violation.of("Expense amount is greater than the max amount"));
        }

        return Optional.empty();
    }
}

File: LLD/InterviewExamples/RuleEngine/RuleEngineNode/lib/services/rules/impl/MaximumAmountRule.js
import Rule  from '../Rule.js'
import Violation from '../Violation.js'
export default class MaximumAmountRule extends Rule{
    constructor(maxAmount) {
        super()
        this.maxAmount = maxAmount
    }

    check(expense) {
        if(Array.isArray(expense)) {
            let currAmount = 0
            for(let e of expense) {
                currAmount += e.getAmountUsd
                if(currAmount > this.maxAmount) {
                    return new Violation("Expense amount exceeds trip max amount limit of "+this.maxAmount)
                }
            }
        } else {
            if(expense.getAmountUsd() > this.maxAmount) {
                return new Violation("Expense amount exceeds max amount limit of "+this.maxAmount)
            }
        }

        return null
    }
}